### PR TITLE
Using an inmemory heartbeats cache to avoid so much contention with saving docs into raven

### PR DESCRIPTION
**Pull request by johnsimons**
This addresses the issue raised in #639

### CODE REVIEW

**Comment by johnsimons** [comment]
@mikeminutillo is this what u have in mind ?

**Comment by johnsimons** [comment]
Also I wonder if it would be better to use the patch api to do it all ?

**Comment by johnsimons** [comment]
After reviewing the changes, we cannot use the patch api because based on certain state changes events get published

**Comment by johnsimons** [comment]
@mikeminutillo this one is done, can u review

**Comment by johnsimons** [comment]
So I have hardcoded 2 minutes for the sync timer, this means it can take up to 2 minutes for an endpoint to be seen as dead or alive, Is this ok ?

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs [hunk 1]
```diff
@@ -147,5 +149,114 @@ void Initialise()
...
+                            };
+
+                            Bus.Publish(new HeartbeatingEndpointDetected
+                            {
+                                Endpoint = heartbeat.EndpointDetails,
+                                DetectedAt = heartbeat.LastReportAt,
+                            });
+                        }
+
+                        session.Store(heartbeat);
```
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: Should only need to do this for new heartbeats. If this is an existing heartbeat, it is already tracked as a part of the session.

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs [hunk 2]
```diff
@@ -147,5 +149,114 @@ void Initialise()
...
+
+                        HeartbeatStatusProvider.RegisterHeartbeatingEndpoint(heartbeat.EndpointDetails, heartbeat.LastReportAt);
```
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: Shouldn't need to do this. We just pulled the info from the heartbeat stats provider

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs [hunk 3]
```diff
@@ -147,5 +149,114 @@ void Initialise()
...
+                }
+            }
+
+            void Refresh(object _)
+            {
+                Sync();
+
+                try
+                {
+                    timer.Change(TimeSpan.FromMinutes(2), Timeout.InfiniteTimeSpan);
```
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: Does this mean when an endpoint starts beating again that we won't notify anyone for 2 minutes?

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs [hunk 4]
```diff
@@ -56,7 +58,7 @@ void Refresh(object _)
 
                 try
                 {
-                    timer.Change((int)TimeSpan.FromSeconds(5).TotalMilliseconds, -1);
+                    timer.Change(TimeSpan.FromSeconds(5), Timeout.InfiniteTimeSpan);
```
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: Should we use [`PeriodicExecutor`](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Infrastructure/PeriodicExecutor.cs) in these startup tasks?

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs (hunk 3 above)
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: Same for new heartbeats

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs [hunk 5]
```diff
@@ -147,5 +149,114 @@ void Initialise()
...
+
+                    foreach (var memHeartbeat in inmemoryHeartbeats)
+                    {
+                        Heartbeat heartbeat;
+                        if (storedHeartbeats.TryGetValue(memHeartbeat.Key, out heartbeat))
+                        {
+                            if (heartbeat.Disabled)
+                            {
+                                continue;
+                            }
```
Comment (mikeminutillo) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: This means that we don't update the details of heartbeats when we aren't monitoring them. Were we before? Should we?

**Comment by mikeminutillo** [comment]
> So I have hardcoded 2 minutes for the sync timer, this means it can take up to 2 minutes for an endpoint to be seen as dead or alive, Is this ok ?

I think it is not OK. One of the discussions we had around heartbeats is that they need to be low latency.
@Particular/servicecontrol-maintainers please chime in.

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs (hunk 5 above)
Comment (johnsimons) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: That is how we were doing it before, not sure if it makes sense updating the last heartbeat timestamp.
If we look at the whole reason to stop monitoring an endpoint, most times has to do with decomissioning an endpoint and hence that endpoint will not be sending heartbeats anyway, so I don't think we need to worry.

File: src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs (hunk 4 above)
Comment (johnsimons) [inline src/ServiceControl/HeartbeatMonitoring/HeartbeatsMonitor.cs]: PeriodicExecuter uses Tasks + Task.Delay to in essence replicate what a Timer gives us for free.
In other words, I am not sure we should be using it at all

**Comment by johnsimons** [comment]
@mikeminutillo did u endup working on this one ?

**Comment by mikeminutillo** [comment]
@johnsimons yeah but I've stashed it at the moment. Let me pull it out and see where I got up to

**Pull request closed without merging**